### ISSUE TITLE
Submit-Renewal - Date casting error

### ISSUE THREAD

**Issue by Freeben666** [rmbolger/Posh-ACME#578]
Hello,

I'm having some trouble with Posh-ACME when trying to renew a certificate:
![image](https://github.com/user-attachments/assets/a8d0b736-25f1-4dd0-8091-b064ac7560fe)

```
MethodInvocationException: C:\Users\XXXXXX\Documents\PowerShell\Modules\Posh-ACME\4.25.1\Private\Update-PAOrder.ps1:94
Line |
  94 |  …         if ((Get-DateTimeOffsetNow) -gt [DateTimeOffset]::Parse($rene …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling "Parse" with "1" argument(s): "String '12/27/2024 09:45:40' was not recognized as a valid DateTime."
MethodInvocationException: C:\Users\XXXXXX\Documents\PowerShell\Modules\Posh-ACME\4.25.1\Public\Submit-Renewal.ps1:51
Line |
  51 |  …         if (-not $Force -and $null -ne $order.RenewAfter -and (Get-Da …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling "Parse" with "1" argument(s): "String '12/27/2024 09:45:40' was not recognized as a valid DateTime."
MethodInvocationException: C:\Users\XXXXXX\Documents\PowerShell\Modules\Posh-ACME\4.25.1\Private\Update-PAOrder.ps1:94
Line |
  94 |  …         if ((Get-DateTimeOffsetNow) -gt [DateTimeOffset]::Parse($rene …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling "Parse" with "1" argument(s): "String '12/27/2024 09:45:40' was not recognized as a valid DateTime."
WARNING: This certificate order has already been completed. Use -Force to overwrite the current certificate.
```

Could that be related to regional settings?

**Comment by rmbolger** [comment]
Hi @Freeben666, thanks for reaching out. I have a feeling this partially a regional issue and partially a bug in the new ARI implementation where I didn't account for an [old bug/feature](https://github.com/PowerShell/PowerShell/issues/13592) in PS 6+ that changed how the native JSON parser handles dates since.

The date/time string that the error is saying it can't parse isn't the right format for what it should be parsing at those lines of code. It should be an ISO 8601/RFC3339 formatted date/time which is what the ACME server should be returning. But if PowerShell is converting it to a DateTime object automatically there might be an implicit string conversion that's messing things up. Can you do me a favor and run the following test and post the output?

```powershell
# if you have multiple orders defined, make sure one is selected with Set-PAOrder that
# has a non-expired cert associated with it
$ariurl = '{0}/{1}' -f (Get-PAServer).renewalInfo,(Get-PACertificate).ARIId
(Invoke-RestMethod $ariurl).suggestedWindow | Get-Member
```

The output should show something like this where the `start` and `end` NoteProperty fields are strings formatted as ISO 8601.
```
   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
end         NoteProperty string end=2024-12-27T12:04:06.666666667Z
start       NoteProperty string start=2024-12-25T12:04:06.666666667Z
```

**Comment by Freeben666** [comment]
Hello Ryan,

Here's the output on the system where I first encountered the bug:
![image](https://github.com/user-attachments/assets/4a8ffadf-8132-4e2e-9d72-5dff8d9429db)

```
Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
end         NoteProperty datetime end=29/12/2024 09:45:40
start       NoteProperty datetime start=27/12/2024 09:45:40
```

The regional settings on that machine are set to `French (France)`.

**Comment by rmbolger** [comment]
Excellent. Thanks! That at least confirms what I thought might be the problem. What's weird is that I can't actually reproduce it on my machine. My local PowerShell 7.4.6 instance should be doing the same thing (albeit with US region formatting), but it's not and I can't figure out why. But that's a personal problem I can deal with later.

In any case, the fix should be relatively simple and I'll try to get it out soon. Until then, you can disable the bugged ARI functionality like this:
```powershell
Set-PAServer -DisableARI
```

However, if the non-ISO date string got saved to the order JSON on disk, things might still end up throwing errors unless you either manually tweak the JSON or just delete and re-create the order from scratch.

**Comment by Freeben666** [comment]
I don't know if that helps, but here's the environment in which I encountered the bug:

- Windows Server 2016 with latest updates
- Microsoft PowerShell 7.4.6 freshly downloaded from the releases of the official Github depot
- Regional settings set to French (France)
- Only modules loaded (or even installed) are Posh-ACME and Posh-ACME.Deploy (and IISAdministration, because PA.Deploy asked for it)

That's all I can think of that could have an impact.

I'll try the -DisableARI modifier, thanks :-)

**Comment by rmbolger** [comment]
I think I figured out why I never ran into the issue on my machine. The auto-conversion to DateTime seems to be skipped if the ISO-date has more than 7 digits in the fractional seconds spot. Let's Encrypt's ACME server seems to often (but not always) give responses with values that have 9 digits. Check out the output difference between these two:

```
PS > @{start='2024-12-15T04:40:12.12345678Z'} | convertto-json | convertfrom-json | gm

   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
start       NoteProperty string start=2024-12-15T04:40:12.12345678Z

PS > @{start='2024-12-15T04:40:12.1234567Z'} | convertto-json | convertfrom-json | gm

   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
start       NoteProperty datetime start=12/15/2024 4:40:12 AM
```

It doesn't affect the fix. I just wanted to solve the mystery on my machines.

**Comment by rmbolger** [comment]
This is now fixed in 4.26.0

**Comment by Freeben666** [comment]
> I think I figured out why I never ran into the issue on my machine. The auto-conversion to DateTime seems to be skipped if the ISO-date has more than 7 digits in the fractional seconds spot. Let's Encrypt's ACME server seems to often (but not always) give responses with values that have 9 digits. Check out the output difference between these two:
> 
> ```
> PS > @{start='2024-12-15T04:40:12.12345678Z'} | convertto-json | convertfrom-json | gm
> 
>    TypeName: System.Management.Automation.PSCustomObject
> 
> Name        MemberType   Definition
> ----        ----------   ----------
> Equals      Method       bool Equals(System.Object obj)
> GetHashCode Method       int GetHashCode()
> GetType     Method       type GetType()
> ToString    Method       string ToString()
> start       NoteProperty string start=2024-12-15T04:40:12.12345678Z
> 
> PS > @{start='2024-12-15T04:40:12.1234567Z'} | convertto-json | convertfrom-json | gm
> 
>    TypeName: System.Management.Automation.PSCustomObject
> 
> Name        MemberType   Definition
> ----        ----------   ----------
> Equals      Method       bool Equals(System.Object obj)
> GetHashCode Method       int GetHashCode()
> GetType     Method       type GetType()
> ToString    Method       string ToString()
> start       NoteProperty datetime start=12/15/2024 4:40:12 AM
> ```
> 
> It doesn't affect the fix. I just wanted to solve the mystery on my machines.

Nice catch !



> This is now fixed in 4.26.0

Thank you very much ☺️
